### PR TITLE
Fix:    [Bug]: triplot with 'ls' argument yields TypeError #21995

### PR DESCRIPTION
**Pull request by gajendra0180**
> The work here:
> 
> Task1 : add that machinery to triplot
> Task2 : add a test (probably with check_figures_equal and pass in the line style via both ls and linestyle)

Modified the triplot file and it is working fine now.
**Task1 Done.**

In Task2,
> and pass in the line style via both ls and linestyle pass in the line style via both ls and linestyle

Made the test file as _test_check_figures_equal.py_ and added some code for both ls and linestyle-
**Task2 Done.**

Attaching some screenshots for the same.

 **Reproduced the bug:**
Reproduced the error

**Fixed the bug:**
Fixed the bug

**The _test_check_figures_equal.py_ file**
test_check_figures_equal

**testing with _test_check_figures_equal.py_ file**
test_passed

Thanks.

### CODE REVIEW

File: lib/matplotlib/tests/test_check_figures_equal.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+
+
+import matplotlib.pyplot as plt
+
+
+def test_check_figures_equal():
```
Comment (timhoffm) [inline lib/matplotlib/tests/test_check_figures_equal.py:6]: You should use the check_figures_equal decorator. See

https://matplotlib.org/devdocs/api/testing_api.html?highlight=check_figures_equal#matplotlib.testing.decorators.check_figures_equal

Comment (gajendra0180) [reply lib/matplotlib/tests/test_check_figures_equal.py]: Hello, I have used the _check_figures_equal_ decorator in the test file.
Please review it.
Thanks.

File: lib/matplotlib/tri/triplot.py [hunk 2]
```diff
@@ -42,6 +44,7 @@ def triplot(ax, *args, **kwargs):
     linestyle, marker, color = matplotlib.axes._base._process_plot_format(fmt)
 
     # Insert plot format string into a copy of kwargs (kwargs values prevail).
+    kwargs = cbook.normalize_kwargs(kwargs, mlines.Line2D)
     kw = kwargs.copy()
```
Comment (maweigert) [inline lib/matplotlib/tri/triplot.py:48]: Just a small comment: As `cbook.normalize_kwargs` already creates a new dictionary, this additional copy might be superfluous and thus not necessary anymore.

Comment (gajendra0180) [reply lib/matplotlib/tri/triplot.py]: Thanks. Made the Required Changes. Please Review.

Comment (gajendra0180) [reply lib/matplotlib/tri/triplot.py]: Thanks.
Made the required changes. 
Please Review.

**Comment by gajendra0180** [comment]
@tacaswell Please Review.
Thanks.

File: lib/matplotlib/tests/test_triplot_with_ls.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from matplotlib.testing.decorators import check_figures_equal
```
Comment (jklymak) [inline lib/matplotlib/tests/test_triplot_with_ls.py:1]: We don't typically have one-issue test files.  Please find where triplot is tested and move there.  Thanks!

Comment (gajendra0180) [reply lib/matplotlib/tests/test_triplot_with_ls.py]: Moved the test to _lib/matplotlib/tests/test_triangulation.py_ test file and improved formatting.
Please Review.
Thanks.

File: lib/matplotlib/tests/test_triangulation.py [hunk 4]
```diff
@@ -19,9 +19,9 @@ def test_delaunay():
     x, y = np.meshgrid(np.linspace(0.0, 1.0, nx), np.linspace(0.0, 1.0, ny))
     x = x.ravel()
     y = y.ravel()
-    npoints = nx*ny
-    ntriangles = 2 * (nx-1) * (ny-1)
-    nedges = 3*nx*ny - 2*nx - 2*ny + 1
+    npoints = nx * ny
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_triangulation.py:22]: Please don't make unrelated changes to the file.

File: lib/matplotlib/tri/triplot.py [hunk 5]
```diff
@@ -42,11 +44,11 @@ def triplot(ax, *args, **kwargs):
     linestyle, marker, color = matplotlib.axes._base._process_plot_format(fmt)
 
     # Insert plot format string into a copy of kwargs (kwargs values prevail).
-    kw = kwargs.copy()
+    kw = cbook.normalize_kwargs(kwargs, mlines.Line2D)
     for key, val in zip(('linestyle', 'marker', 'color'),
                         (linestyle, marker, color)):
         if val is not None:
-            kw[key] = kwargs.get(key, val)
+            kw[key] = kw.get(key, val)
```
Comment (QuLogic) [inline lib/matplotlib/tri/triplot.py:51]: What this is really saying is:
```suggestion
            kw.setdefault(key, val)
```

File: lib/matplotlib/tests/test_triangulation.py (hunk 4 above)
Comment (gajendra0180) [reply lib/matplotlib/tests/test_triangulation.py]: Yes, Sorry for that.

File: lib/matplotlib/tri/triplot.py (hunk 5 above)
Comment (gajendra0180) [reply lib/matplotlib/tri/triplot.py]: Thanks. Made the changes.

File: lib/matplotlib/tests/test_triangulation.py [hunk 6]
```diff
@@ -1177,3 +1177,10 @@ def test_tricontourset_reuse():
     assert tcs2._contour_generator != tcs1._contour_generator
     tcs3 = ax.tricontour(tcs1, z)
     assert tcs3._contour_generator == tcs1._contour_generator
+
+
+@check_figures_equal()
+def test_triplot_with_ls(fig_test, fig_ref):
+    fig_test.subplots().triplot([0, 2, 1], [0, 0, 1], [[0, 1, 2]], ls='--')
+    fig_ref.subplots().triplot([0, 2, 1], [0, 0, 1], [
+        [0, 1, 2]], linestyle='--')
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_triangulation.py:1186]: I would re-factor the data out so it's clear that it's the same (and also no wrapping would be needed.)

Comment (gajendra0180) [reply lib/matplotlib/tests/test_triangulation.py:1186]: Thanks .
I didn't actually get what you meant, do I need to do any changes or anything like that?

Comment (QuLogic) [reply lib/matplotlib/tests/test_triangulation.py:1186]: Something like
```suggestion
    x = [0, 2, 1]
    y = [0, 0, 1]
    data = [[0, 1, 2]]
    fig_test.subplots().triplot(x, y, data, ls='--')
    fig_ref.subplots().triplot(x, y, data, linestyle='--')
```

Comment (gajendra0180) [reply lib/matplotlib/tests/test_triangulation.py]: Thanks. 👍 Made the changes .I will take care of it from now onwards.

File: lib/matplotlib/tri/triplot.py [hunk 7]
```diff
@@ -42,11 +44,11 @@ def triplot(ax, *args, **kwargs):
     linestyle, marker, color = matplotlib.axes._base._process_plot_format(fmt)
 
     # Insert plot format string into a copy of kwargs (kwargs values prevail).
-    kw = kwargs.copy()
+    kw = cbook.normalize_kwargs(kwargs, mlines.Line2D)
     for key, val in zip(('linestyle', 'marker', 'color'),
                         (linestyle, marker, color)):
         if val is not None:
-            kw[key] = kwargs.get(key, val)
+            kw.setdefault(key, val)
```
Comment (timhoffm) [inline lib/matplotlib/tri/triplot.py:51]: I think this changes precedence if both are given, a linestyle via fmt and a linestyle via kwarg.

Comment (QuLogic) [reply lib/matplotlib/tri/triplot.py:51]: `kwargs.get(key, val)` gets the value for `key`, falling back to `val` if it doesn't exist, then setting that on the same key in `kw[key]`. `kw.setdefault(key, val)` sets `key` to `val` if it's not already set. They should be the same. Note, `kw` is a copy of `kwargs`.

Comment (timhoffm) [reply lib/matplotlib/tri/triplot.py:51]: You‘re right.

**Review by timhoffm — APPROVED** [review]
This must be squashed into a single commit to satisfy the PR cleanliness check. -can be done either by the author or by the second reviewer on merge.

**Comment by tacaswell** [comment]
Thank you @gajendra0180 !  Congratulations on your first merged Matplotlib PR 🎉 , hopefully we will hear from you agian.
